k8s/{client,resource}: API improvements and support for custom retries

joamaki opened the pull request:
This PR includes multiple small improvements to pkg/k8s/resource and pkg/k8s/client libraries found during review of https://github.com/cilium/cilium/pull/21617. 

*logging: Initialize klog from SetupLogging*
This delays the klog bridging to logrus so we don't have logging related goroutines running for tests and we can use the goleak library to check for leaked goroutines.

*k8s/client: Set up clientset in constructor*
It's useful to be able to use the clientset object before starting to get a narrower interface from the clientset for constructing e.g. ListerWatcher: `clientset.CoreV1().Pods("")` does not perform actual API calls so should be allowed.

*k8s/resource: Make retries configurable*
Adds configurable retries to resource and simplifies the construction of the resource:
`func New[T k8sRuntime.Object](lc hive.Lifecycle, lw cache.ListerWatcher, opts ...Option) Resource[T]`
This commit also fixes a goroutine leak as the DelayingQueue in client-go's workqueue didn't implement
override for ShutDownWithDrain. Workaround is to call both ShutDownWithDrain() and ShutDown().

*hive: Make Hook into an interface*
Often the lifecycle hook is implemented by an object. This changes `Append` to take an interface instead of the hook struct
to simplify use in the common case. Downside is that we need special handling for the hook struct to show the correct function name and to support optionality (OnStart=nil).

*stream: Add goroutine leak checks*
Add the goleak check to the stream library and fix an issue in the Debounce operator: the errors channel given to ToChannel was not drained on context cancellation, leading to ToChannel's goroutine getting blocked on send to error channel.

*k8s/resource: Drop Store[T] from SyncEvent*
Drop the Store[T] from the sync event. It's already available via Store() and usually not useful.

*k8s/resource: Make handlers optional in Event.Handle()*
Allow `event.Handle(nil, onUpdate, onDelete)` as quite often we don't need to handle the sync event.

joamaki committed:
k8s/client: Set up clientset in constructor

The clientset was only constructed from OnStart in order to stop use of it prior
to it being started (so that we can inspect the object graph without unwanted
side-effects). This however made it inconvenient to use in cases where a specific
interface was being extracted from the clientset, e.g. to build a ListerWatcher.

The usability hit is not worth it, so change the k8s client to construct a
usable clientset at construction time.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s/resource: Make retries configurable

Instead of a fixed 5 retries allow constructing the resource
with a configurable retry strategy:

    type ErrorHandler func(key Key, numRetries int, err error) ErrorAction
    func WithErrorHandler(h ErrorHandler) Option

Resource is now constructed with New that takes functional options:

    func New[T k8sRuntime.Object](lc hive.Lifecycle, lw cache.ListerWatcher, opts ...Option) Resource[T]

This new API now allows building the ListerWatcher using other objects:

    cell.Provide(
      func(lc hive.Lifecycle, c client.Clientset, cfg MyConfig) resource.Resource[*MyObject] {
        lw := /* constructed based on 'cfg' */
        return resource.New(lc, lw)
      })

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
logging: Initialize klog from SetupLogging

The init() in logging bridges the klog library (used by client-go)
to logrus. It uses the WriterLevel() method which forks of a goroutine
that implements a io.Writer using a pipe.

This unfortunately leaves multiple goroutine around that causes false
positives for goroutine leak checks. Since SetupLogging() is always
called in our applications we can perform the klog initialization
from there and allow tests to do goroutine checks.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
hive: Make Hook into an interface

Simplifies dealing with lifecycle in the normal case:

  func (*MyObject) Start(context.Context) error { ... }
  func (*MyObject) Stop(context.Context) error { ... }
  func New(lc fx.Lifecycle) *MyObject {
    mo := &MyObject{}
    lc.Append(mo)
    return mo
  }

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
stream: Add goroutine leak checks

As stream is another library besides pkg/k8s/resource that uses a lot of goroutines,
add the leak checks here as well.

Fix an issue in Debounce. By handling the context cancellation directly instead of
relying on error from the src observable, Debounce ended up not draining errors
channel and causing a goroutine leak (ToChannel blocked on sending to errs channel).

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink commented on pkg/k8s/resource/resource.go:
```diff
@@ -46,19 +45,18 @@ type Resource[T k8sRuntime.Object] interface {
+// New creates a new Resource[T]. Use with hive.Provide:
 //
-//	var exampleCell = cell.Module(
+//	var exampleCell = hive.Module(
```
`Module` and `Provide` are still located in the cell package.

joamaki replied:
Ah right, fixed!

joamaki committed:
k8s/resource: Drop Store[T] from SyncEvent

The store is already available via Resource.Store() and in most cases
one uses either the Store directly, or the stream of events. When using
the event stream the store is usually not needed as subscribers build
their own projected view of the state via update events.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

squeed commented on pkg/k8s/resource/example/main.go:
```diff
@@ -54,16 +53,20 @@ func main() {
-				return utils.ListerWatcherFromTyped[*corev1.ServiceList](c.CoreV1().Services(""))
-			},
-		),
+		func(lc hive.Lifecycle, c client.Clientset) resource.Resource[*corev1.Pod] {
```
I'm debating if this is too much boilerplate, or if the alternative would be brain-melting amounts of reflection. Hm.

dylandreimerink replied:
Hopefully, future go versions are better at inferring type parameters so we don't have to repeat them all over the place.

joamaki replied:
Yep I've been debating back and forth about the boilerplate. The alternative is:

```
cell.Decorate(
  func(c client.Clientset, some A, other B, stuff C) cache.ListerWatcher {
    // make the ListerWatcher,
  },
  resource.New[*corev1.Pod], // depends on cache.ListerWatcher
)
```

But in the end I thought the indirection here is not worth it. Maybe if at some point `Resource` depends on many many more things we'll want to do it somewhat along these lines.

squeed replied:
Yeah, I agree - leave it for now and see where the pain points are.

joamaki committed:
k8s/resource: Make handlers optional in Event.Handle()

In many cases the subscriber to resource events does not care to handle
some event types (e.g. sync). Make the handler functions optional so
the subscriber doesn't need to implement a no-op handler.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

squeed commented on pkg/k8s/resource/resource.go:
```diff
@@ -76,28 +74,23 @@ type Resource[T k8sRuntime.Object] interface {
+	for _, applyOpt := range opts {
+		applyOpt(&r.opts)
+	}
+	r.ctx, r.cancel = context.WithCancel(context.Background())
```
What is this context used for? Shouldn't it be provided by `OnStart()`?

joamaki replied:
No the ctx in start is only valid for the duration of the start itself. It is used to cancel start if things take too long to setup. E.g. say a database module can block on start to connect to a database and if it takes too long the start can be aborted by cancelling the start context.

This context here is for the whole duration of the resource and is cancelled when stopping.

I’ve for the whole time I’ve worked with fx/dig been wondering whether we should just remove the context from the start as it is confusing. The alternative is to not have it and then just ”fatal” if start gets stuck. We wouldn’t have the opportunity to then clean up things, but we haven’t needed that thus far anyway…

squeed replied:
Oof, that's very, very confusing. I would assume that `Start(ctx)` receives the "app-level" context, not the "startup-level" context. I understand the utility of having separate "hook" contexts vs. "app" contexts, especially for shutdown.

I suppose you could catch applications that misbehave by immediately cancelling the Start context after all hooks are done. But perhaps better is just to wire-in an app-level or hive-level context?

Honestly, a stop context makes some sense to me, but not a start context.

joamaki replied:
Agreed, useful but confusing. Start context is cancelled after start has completed, so we do catch misuses. We could consider using a new type for the start and stop contexts? e.g. `Start(hive.StartTimeoutContext)`. That should draw attention to what it is and still allow using it as `context.Context` when needed. WDYT?

I would not do app-level contexts at all in the agent, but rather only use context locally within modules when it makes sense. One reason for this is that a global context encourages building cleanup around it, but that's almost never enough as you also need to wait for things to actually finish (e.g. background goroutines to go away and so on. Especially important in tests to not leave stuff around). Another problem is that there would be no ordering in when things stop. We might end up stopping something that some other module is still using!

joamaki replied:
Thought about this a bit. I really don't like the idea of removing the start context as there's plenty of places where we do blocking operations from the start hook: for example the k8s-client blocks until it actually can connect to the api-server, local node info is initialized at start from api-server and so on. We really should handle these sort of timeouts gracefully and provide clean reasons in the logs what went wrong, e.g. "this start hook timed out" rather than Fatal with non-specific error message + dump stack trace, and to give the possibility to gracefully stop modules started before it.

So I've now added `hive.HookContext` as a new type of context.Context and documented its purpose. Hope that clears up the confusion in the future.

joamaki committed:
hive: Use a new type for the start and stop contexts to avoid confusion

The context passed to Start and Stop hooks is used to abort the start and
stop in case of a timeout. This however is not that evident to devs new to
this.

This changes context.Context to hive.HookContext to make it clear that
the context has a narrow purpose.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joestringer commented on pkg/k8s/resource/example/main.go:
```diff
@@ -54,16 +53,20 @@ func main() {
+			return resource.New[*corev1.Pod](lc, lw)
+		},
+		func(lc hive.Lifecycle, c client.Clientset) resource.Resource[*corev1.Service] {
+			if !c.IsEnabled() {
+				return nil
+			}
```
nit, I'm curious why these functions should check if the clientset is enabled or not and return nil for that case. Why doesn the caller do this check? Is this because we need to register the resource whether the clientset is enabled or not, eg to enable the functionality where we print the tree of cells?

joamaki replied:
If the client is not enabled then all the embedded clientsets in `Clientset` are nil (e.g. clientset.CoreV1() is a nil deref), so we need to check for that. I guess alternatively we could return Clientset as `nil` instead of having a "disabled" clientset. I think I was just replicating the `k8s.IsEnabled` without thinking too much.

And yes, I'm choosing here to have a "static" object graph that is not conditional on whether or not features are enabled by config. I think that's the safer approach as then we'll need to test only once that all dependencies are satisfied, and we can produce one dot graph visualization that includes everything instead of gazillion combinations.

We could build this in a way to only `Provide` when `Config.Enable` is true, but this would then require all dependees to use `cell.In` struct with `optional:"true"` struct tags and check if the field is nil.

joestringer commented on pkg/stream/operators.go:
```diff
@@ -193,6 +195,7 @@ func Debounce[T any](src Observable[T], duration time.Duration) Observable[T] {
 					select {
 					case <-done:
 						complete(ctx.Err())
+						<-errs
```
Isn't this throwing away the errors? Should those be logged if non-nil? Or is it not possible to have errors in the `<-done` case?

joamaki replied:
Ah good point. I made this overly complicated. I don't need this case at all since `errs` will get the "context canceled" error from `src` anyway. Will fix.

joestringer commented on pkg/k8s/resource/event.go:
```diff
@@ -10,15 +10,18 @@ import (
+	// Handle the event by invoking the right handler. The handlers can be
+	// nil in which case the event is just marked as processed successfully.
+	//
+	// On errorr from the event handlers the event is requeued by key for later retrying.
```
nit:
```suggestion
	// On error from the event handlers the event is requeued by key for later retrying.
```

joamaki replied:
Fixed.

joestringer reviewed (approved):
LGTM, though most of this code is pretty much brand new to me. I'd put more emphasis on others' reviews.

squeed reviewed (approved):
Looks good, thanks!

qmonnet reviewed (approved):
Thanks!
